File: workspace/workflows/src/metadata/transition_to.rs
use crate::activity::{Activity, ActivityContext, Error};
use async_trait::async_trait;
use serde_json::Value;
use bosca_client::client::{Client, WorkflowJob};
use bosca_client::client::add_activity::ActivityInput;

pub struct MetadataTransitionToActivity {
    id: String,
}

impl Default for MetadataTransitionToActivity {
    fn default() -> Self {
        Self::new()
    }
}

impl MetadataTransitionToActivity {
    pub fn new() -> MetadataTransitionToActivity {
        MetadataTransitionToActivity {
            id: "metadata.transition.to".to_string(),
        }
    }
}

#[async_trait]
impl Activity for MetadataTransitionToActivity {
    fn id(&self) -> &String {
        &self.id
    }

    fn create_activity_input(&self) -> ActivityInput {
        let mut configuration = serde_json::Map::new();
        configuration.insert("state".to_string(), Value::String("draft".to_string()));
        configuration.insert("status".to_string(), Value::String("marked draft".to_string()));
        ActivityInput {
            id: self.id.to_owned(),
            name: "Finalize Metadata Transition".to_string(),
            description: "Finalize a Metadata Transition".to_string(),
            child_workflow_id: None,
            configuration: Value::Object(configuration),
            inputs: vec![],
            outputs: vec![],
        }
    }

    async fn execute(&self, client: &Client, _: &mut ActivityContext, job: &WorkflowJob) -> Result<(), Error> {
        let state = job
            .workflow_activity
            .configuration
            .get("state")
            .unwrap()
            .as_str()
            .unwrap()
            .to_string();
        let status_value = job.workflow_activity.configuration.get("status").unwrap();
        let status = if status_value.is_null() {
            "".to_string()
        } else {
            status_value.as_str().unwrap().to_string()
        };
        let metadata = &job.metadata.clone().unwrap();
        client
            .set_workflow_state_complete(&metadata.id, &status)
            .await?;
        client
            .set_workflow_state(&metadata.id, &state, &status, true)
            .await?;
        Ok(())
    }
}

File: workspace/runners/kotlin/activities/src/main/kotlin/io/bosca/workflow/metadata/TransitionTo.kt
package io.bosca.workflow.metadata

import io.bosca.api.Client
import io.bosca.graphql.fragment.WorkflowJob
import io.bosca.graphql.type.ActivityInput
import io.bosca.util.toOptional
import io.bosca.workflow.Activity
import io.bosca.workflow.ActivityContext

class TransitionTo(client: Client) : Activity(client) {

    override val id: String = ID

    override suspend fun toActivityDefinition(): ActivityInput {
        return ActivityInput(
            id = id,
            name = "Finalize Metadata Transition",
            description = "Finalize a Metadata Transition",
            configuration = mapOf<String, Any>(
                "state" to "draft",
                "status" to "marked draft",
            ).toOptional(),
            inputs = emptyList(),
            outputs = emptyList(),
        )
    }

    override suspend fun execute(context: ActivityContext, job: WorkflowJob) {
        val configuration = job.workflowActivity.workflowActivity.configuration as Map<*, *>
        val state = configuration["state"] as String
        val status = if (configuration.containsKey("status")) configuration["status"] as String else ""
        client.workflows.setMetadataWorkflowStateComplete(job.metadata?.metadata?.id ?: error("missing metadata id"), status)
        client.workflows.setMetadataWorkflowState(job.metadata!!.metadata.id, state, status, true)
    }

    companion object {
        const val ID = "metadata.transition.to"
    }
}